build/modules/gke/cluster.tf

```diff
@@ -0,0 +1,155 @@
+# Copyright 2019 Google LLC All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+provider "google-beta" {
+  version = "~> 2.4"
+  zone      = "${lookup(var.cluster, "zone")}"
+}
+
+
+provider "google" {
+  version = "~> 2.4"
+}
+
+data "google_client_config" "default" {}
+
+# echo command used for debugging purpose
+# Run `terraform taint null_resource.test-setting-variables` before second execution
+resource "null_resource" "test-setting-variables" {
+    provisioner "local-exec" {
+      command = "${"${format("echo Current variables set as following - name: %s, project: %s, machineType: %s, initialNodeCount: %s, zone: %s, legacyAbac: %s",
+      "${lookup(var.cluster, "name")}", "${lookup(var.cluster, "project")}",
+      "${lookup(var.cluster, "machineType")}", "${lookup(var.cluster, "initialNodeCount")}",
+      "${lookup(var.cluster, "zone")}", "${lookup(var.cluster, "legacyAbac")}")}"}"
+    }
+}
+
+
+locals {
+  username = "${var.password != "" ? var.username : ""}"
+}
+
+# assert that password has correct length
+# before creating the cluster to avoid 
+# unfinished configurations
+resource "null_resource" "check-password-length" {
+  count = "${length(var.password) >= 16 || length(var.password) == 0 ? 0 : 1}"
+  "Password must be more than 16 chars in length" = true
+}
+
+resource "google_container_cluster" "primary" {
+  name     = "${lookup(var.cluster, "name")}"
+  location     = "${lookup(var.cluster, "zone")}"
+  project  = "${lookup(var.cluster, "project")}"
+  provider = "google-beta"
+
+  # Setting an empty username and password explicitly disables basic auth
+  master_auth {
+    username = "${local.username}"
+    password = "${var.password}"
+  }
+  enable_legacy_abac = "${lookup(var.cluster, "legacyAbac")}"
+  node_pool = [
```

Comment:
After reading the docs about how terraform handles updates to node pools, I think it would be better to make each node pool a separate resource.

From https://www.terraform.io/docs/providers/google/r/container_cluster.html#node_pool:

> **Warning**: node pools defined inside a cluster can't be changed (or added/removed) after cluster creation without deleting and recreating the entire cluster. Unless you absolutely need the ability to say "these are the *only* node pools associated with this cluster", use the [google_container_node_pool](https://www.terraform.io/docs/providers/google/r/container_node_pool.html) resource instead of this property.

Reply:
That would make cluster creating longer, about 10 minutes compared to 3 minutes now.
But regarding this warning it seems to be inevitable.

Reply:
Note that after this change I faced issue with cluster starts upgrading after I attach 2 additional node_pools to it.
https://github.com/terraform-providers/terraform-provider-google/issues/3385

Reply:
Fixed the issue as proposed in link I posted above:
  remove_default_node_pool = true
And now we have 3 node pools and helm would install Agones right after creation of all node pools.

Reply:
I agree that cluster creation is much slower. It's really annoying that the terraform implementation for GKE requires jumping through these hoops. :/